### PR TITLE
refactor(array): use `DataType` to replace `ArrayMeta`

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Learning from Arrow, this PR provides a method to [get data type from array](https://docs.rs/arrow-array/latest/arrow_array/array/trait.Array.html#method.data_type). Therefore, `ArrayMeta` can be removed.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

### CODE REVIEW

**Comment by wangrunji0408** [comment]
Oh I realized an old problem. Both int64 and timestamptz type share the I64Array. Maybe it's time to introduce `Timestamptz` type? 🥵 cc @xiangjinwu

**Comment by xxchan** [comment]
> Oh I realized an old problem. Both int64 and timestamptz type share the I64Array. Maybe it's time to introduce `Timestamptz` type?

How's that a problem? & not related with `ArrayMeta`? :eyes:

**Comment by wangrunji0408** [comment]
> How's that a problem? & not related with `ArrayMeta`? :eyes:

Timestamptz values are stored in `I64Array`. Its `data_type()` should be Timestamptz but in this PR it returns Int64.
There's no problem in `ArrayMeta` as both cases return `Simple`. 😇

**Comment by liurenjie1024** [comment]
> Timestamptz values are stored in I64Array. Its data_type() should be Timestamptz but in this PR it returns Int64.
There's no problem in ArrayMeta as both cases return Simple. 😇

In old days we stored data types in array, just like arrow.

**Review by liurenjie1024 — APPROVED** [review]

File: src/common/src/array/primitive_array.rs [hunk 1]
```diff
@@ -48,15 +49,15 @@ where
...
-    /// Creates an `ArrayBuilder` for this primitive type
-    fn create_array_builder(capacity: usize) -> ArrayBuilderImpl;
 
     // item methods
     fn to_protobuf<T: Write>(self, output: &mut T) -> ArrayResult<usize>;
 }
 
 macro_rules! impl_array_methods {
     ($scalar_type:ty, $array_type_pb:ident, $array_impl_variant:ident) => {
+        const DATA_TYPE: DataType = DataType::$array_impl_variant;
```
Comment (xiangjinwu) [inline src/common/src/array/primitive_array.rs:59]: FYI there is one exception that `array_impl_variant`is different from `DataType` variant: `Utf8` vs `Varchar`. But it is not primitive and we are good here.

**Review by xiangjinwu — COMMENTED** [review]
> Oh I realized an old problem. Both int64 and timestamptz type share the I64Array. Maybe it's time to introduce `Timestamptz` type? 🥵 cc @xiangjinwu

This PR lgtm. Just as you have noticed, we would have to introduce a dedicated array type for `timestamptz` before we can safely rely on the `Array::data_type` method introduced here. And I guess it makes sense to have a `Timestamptz(i64)` scalar as well?

Also want to make it clear that there is still no way to get `DataType` from a `Scalar`, in other words by taking scalar out of array we "lose" some type information. Just want to confirm we do not have a motivation now to enrich scalar as well, right?

**Comment by wangrunji0408** [comment]
Sure. We'll introduce `Timestamptz(i64)` scalar and `TimestamptzArray` in the next step. But no plan to enrich scalar.

**Pull request merged**